src/connectionbenchmark/benchmarkdownloadtask.cpp

```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "benchmarkdownloadtask.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "networkrequest.h"
+#include "taskscheduler.h"
+
+#include <QByteArray>
+
+namespace {
+Logger logger(LOG_MAIN, "BenchmarkDownloadTask");
+}
+
+BenchmarkDownloadTask::BenchmarkDownloadTask() : Task("BenchmarkDownloadTask") {
+  MVPN_COUNT_CTOR(BenchmarkDownloadTask);
+}
+
+BenchmarkDownloadTask::~BenchmarkDownloadTask() {
+  MVPN_COUNT_DTOR(BenchmarkDownloadTask);
+}
+
+void BenchmarkDownloadTask::run() {
+  logger.debug() << "Run download";
+
+  m_request = NetworkRequest::createForGetUrl(this, m_fileUrl.toString());
+
+  connect(m_request, &NetworkRequest::requestCompleted, this,
+          [&](const QByteArray& data) {
+            handleTaskFinished(QNetworkReply::NoError, data);
+          });
+  connect(m_request, &NetworkRequest::requestUpdated, this,
+          &BenchmarkDownloadTask::taskProgressed);
+  connect(m_request, &NetworkRequest::requestFailed, this,
+          &BenchmarkDownloadTask::handleTaskFinished);
+
+  m_elapsedTimer.start();
+}
+
+void BenchmarkDownloadTask::stop() {
+  logger.debug() << "Stop download";
+  Q_ASSERT(m_request);
+
+  m_request->abort();
```

Comment:
1. m_request can be null here if the task has not been scheduled yet.
2. we do not have a way to remove a task from the queue yet, so, you probably need a state: "active, inactive". if `stop()` is called before `run()` you must do something.

Reply:
Added `StateCancelled` to cover this case.